### PR TITLE
[ADAG] Add visualization of compiled graphs

### PR DESCRIPTION
**Pull request by Bye-legumes**
## Why are these changes needed?
https://github.com/ray-project/ray/issues/47945

My test case
```
import ray
import time
import random
from ray.dag.input_node import InputNode
from ray.dag.output_node import MultiOutputNode
ray.init(address = "10.218.163.33:6274")
import os
@ray.remote
class Actor:
    def __init__(self, init_value, fail_after=None, sys_exit=False):
        self.i = init_value
        self.fail_after = fail_after
        self.sys_exit = sys_exit

        self.count = 0


    def echo(self, x):
        self.count += 1
        return x



    def sleep(self, x):
        time.sleep(x)
        return x

    @ray.method(num_returns=2)
    def return_two(self, x):
        return x, x + 1

    def read_input(self, x):
        return x




a = Actor.remote(0)
b = Actor.remote(0)
single_fetch = True
with InputNode() as i:
    o1, o2 = a.return_two.bind(i)
    o3 = b.echo.bind(o1)
    o4 = b.echo.bind(o2)
    dag = MultiOutputNode([o3, o4])

compiled_dag = dag.experimental_compile()
for _ in range(3):
    refs = compiled_dag.execute(1)
    if single_fetch:
        for i, ref in enumerate(refs):
            res = ray.get(ref)
            assert res == i + 1
    else:
        res = ray.get(refs)
        assert res == [1, 2]
compiled_dag.visualize()
compiled_dag.teardown()
```
plot
image

## Checks

- [√] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [√] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
cc @kevin85421 can you take a look?

File: python/ray/dag/compiled_dag_node.py [hunk 1]
```diff
@@ -1887,6 +1887,64 @@ async def execute_async(
         self._execution_index += 1
         return fut
 
+    def visualize(self, filename="compiled_dag", format="png", view=False):
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:1890]: compiled_graph

File: python/ray/dag/compiled_dag_node.py [hunk 2]
```diff
@@ -1887,6 +1887,64 @@ async def execute_async(
...
+        """
+        Visualize the compiled DAG using Graphviz.
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:1892]: compiled graph.

Please document when this method can be called, i.e., assumptions of data structure readiness

File: python/ray/dag/compiled_dag_node.py [hunk 3]
```diff
@@ -1887,6 +1887,64 @@ async def execute_async(
...
+                label += type(dag_node).__name__
+
+            # Add the node to the graph
+            dot.node(str(idx), label)
+
+        # Add edges based on downstream tasks
+        for idx, task in self.idx_to_task.items():
+            for downstream_idx, _ in task.downstream_task_idxs.items():
+                # You can also include edge labels with channel types or other info
+                dot.edge(str(idx), str(downstream_idx))
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:1943]: Can you decorate the edges with type hints?

File: python/ray/dag/compiled_dag_node.py [hunk 4]
```diff
@@ -1887,6 +1887,64 @@ async def execute_async(
...
+            elif hasattr(dag_node, "get_method_name"):
+                method_name_attr = getattr(dag_node, "get_method_name")
+                if callable(method_name_attr):
+                    method_name = method_name_attr()
+                else:
+                    method_name = method_name_attr
+
+                # Get actor ID if applicable
+                actor_handle = dag_node._get_actor_handle()
+                if actor_handle:
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:1928]: I see some non-driver nodes don't have actors, why is that?

**Review by ruisearch42 — COMMENTED** [review]
Thanks for the contribution.

File: python/ray/dag/compiled_dag_node.py (hunk 4 above)
Comment (ruisearch42) [reply python/ray/dag/compiled_dag_node.py:1928]: This is due to multi-returns. I think we should make the visualized nodes more understandable by end users.

Comment (ruisearch42) [reply python/ray/dag/compiled_dag_node.py:1928]: One idea is to use rectangles instead of ovals for these special nodes.

**Comment by kevin85421** [comment]
@ruisearch42 is this different from the visualization function in your overlap PR?

File: python/ray/dag/compiled_dag_node.py [hunk 5]
```diff
@@ -1887,6 +1887,64 @@ async def execute_async(
...
+                    actor_id = actor_handle._actor_id.hex()
+                    label += f"Actor: {actor_id[:6]}...\nMethod: {method_name}"
+                else:
+                    label += f"Method: {method_name}"
+            else:
+                label += type(dag_node).__name__
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:1934]: We should explicitly handle `MultiOutputNode` and `ClassMethodNode`, for unexpected types we should highlight (maybe with red color)

File: python/ray/dag/compiled_dag_node.py (hunk 4 above)
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:1928]: This applies to `ClassMethodNode`, there are more specific methods to call, e.g., `is_class_method_output()` to distinguish between different cases, we should use those.

File: python/ray/dag/compiled_dag_node.py [hunk 6]
```diff
@@ -1887,6 +1887,64 @@ async def execute_async(
...
+
+            # Initialize the label
+            label = f"Task {idx}\n"
+
+            # Handle different types of dag_node
+            if isinstance(dag_node, InputNode):
+                label += "InputNode"
+            elif isinstance(dag_node, InputAttributeNode):
+                label += f"InputAttributeNode[{dag_node.key}]"
+            elif hasattr(dag_node, "get_method_name"):
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:1919]: Can you add some unit tests to make sure when internal implementations are changed (say how we handle multi-return), the visualization is not broken? This helps capture such cases early.

**Review by ruisearch42 — COMMENTED** [review]
After addressing these comments, I think we are good to merge!

Over time we can always improve and add more features. This will be very useful.

**Comment by ruisearch42** [comment]
> @ruisearch42 is this different from the visualization function in your overlap PR?

This is to visualize the DAG itself, the overlap PR visualizes the schedule. Both will be useful.

File: python/ray/dag/compiled_dag_node.py (hunk 1 above)
Comment (kevin85421) [inline python/ray/dag/compiled_dag_node.py:1890]: do we call the function anywhere?

File: python/ray/dag/compiled_dag_node.py (hunk 5 above)
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py:1934]: +1

**Review by kevin85421 — COMMENTED** [review]
DAG (non-compiled) seems to have visualization support: https://github.com/ray-project/ray/blob/master/python/ray/dag/vis_utils.py.

What's the difference between the existing one and this PR?

**Comment by Bye-legumes** [comment]
Hi, @ruisearch42 , I just update it and solved you comments. The new plots is above with different color and shape for different node. The tests are also included.

**Comment by Bye-legumes** [comment]
I suspect that graphviz is not available on the test environment so we need  to `pip install graphviz ` and `sudo apt-get install graphviz` for visualization test. Do you know how we can achieve that? I need to add dependency to which file for pip and apt-get? @ruisearch42

File: python/ray/dag/compiled_dag_node.py (hunk 1 above)
Comment (kevin85421) [reply python/ray/dag/compiled_dag_node.py]: I just realized it is an user-facing API.

**Comment by Bye-legumes** [comment]
> > I suspect that graphviz is not available on the test environment so we need to pip install graphviz  and sudo apt-get install graphviz for visualization test.
> 
> @Bye-legumes maybe you can try to use runtime environment.
> 
> Btw, would you mind answering my question here: [#47958 (review)](https://github.com/ray-project/ray/pull/47958#pullrequestreview-2361319963)? Thanks!

This is for the complied graph with different nodes that we have different node handling for different node for complied graph. Also the  data transport method is also included as we have hint. This is the result of that plot.
image

How I can set runtime environment for this test case? Should install packages in the  test.py with `pip install graphviz` and `sudo apt-get install graphviz`

File: python/ray/dag/tests/experimental/test_accelerated_dag.py [hunk 7]
```diff
@@ -2493,6 +2493,91 @@ async def main():
     compiled_dag.teardown()
 
 
+class TestVisualization:
```
Comment (ruisearch42) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:2496]: docstring

File: python/ray/dag/tests/experimental/test_accelerated_dag.py [hunk 8]
```diff
@@ -2493,6 +2493,91 @@ async def main():
...
+            o3 = b.echo.bind(o1)
+            o4 = b.echo.bind(o2)
+            dag = MultiOutputNode([o3, o4])
+
+        compiled_dag = dag.experimental_compile()
+
+        filename = "compiled_graph_multi_return2"
+        compiled_dag.visualize(filename=filename, view=False)
+
+        assert os.path.exists(f"{filename}.png"), "Visualization file was not created."
```
Comment (ruisearch42) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:2572]: We should not be checking the existence of visualization files (which does not give much signal), but the graph structures (nodes and edges)

File: python/ray/dag/tests/experimental/test_accelerated_dag.py
Comment (ruisearch42) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:1]: There are test failures for this file. Please fix.

File: python/ray/dag/compiled_dag_node.py [hunk 9]
```diff
@@ -1887,6 +1887,129 @@ async def execute_async(
...
+    def visualize(self, filename="compiled_graph", format="png", view=False):
+        """
+        Visualize the compiled graph using Graphviz.
+
+        This method generates a graphical representation of the compiled graph,
+        showing tasks and their dependencies.
+
+        **When to Call:**
+        - This method should be called **after** the graph has
+            been compiled using `experimental_compile()`.
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:1899]: Can we add checks? One way is to see if `self.dag_output_channels` is None (maybe there are better approaches)

File: python/ray/dag/compiled_dag_node.py [hunk 10]
```diff
@@ -1887,6 +1887,129 @@ async def execute_async(
...
+        - The internal data structures (`self.idx_to_task`,
+            `self.idx_to_task[task_idx].dag_node`, etc.)
+            must be fully initialized and populated.
+        - All tasks and their relationships (downstream tasks,
+          type hints, etc.) should be ready.
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:1904]: Can we add checks?

File: python/ray/dag/compiled_dag_node.py [hunk 11]
```diff
@@ -1887,6 +1887,129 @@ async def execute_async(
...
+
+        **Assumptions:**
+        - `self.idx_to_task` is a dictionary mapping task
+        indices to `CompiledTask` instances.
+        - Each `CompiledTask` has a `dag_node` attribute that represents the graph node.
+        - Type hints for inputs and outputs are properly set in the graph nodes.
+        - The compiled graph does not contain cycles (it's a DAG).
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:1911]: Are any of these not satisfied after dag.experimental_compile()? maybe we should just remove this.

File: python/ray/dag/compiled_dag_node.py [hunk 12]
```diff
@@ -1887,6 +1887,129 @@ async def execute_async(
...
         return fut
 
+    def visualize(self, filename="compiled_graph", format="png", view=False):
+        """
+        Visualize the compiled graph using Graphviz.
+
+        This method generates a graphical representation of the compiled graph,
+        showing tasks and their dependencies.
+
+        **When to Call:**
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:1897]: I think you can just mention this needs to be called after `experimental_compile()`, and then do actual checking in code

File: python/ray/dag/tests/experimental/test_accelerated_dag.py
Comment (Bye-legumes) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:1]: I suspect that graphviz is not available on the test environment so we need to `pip install graphviz`  and `sudo apt-get install graphviz` for visualization test. Do you know how we can achieve that in  the test environments? I need to add dependency to which file for pip and apt-get? @ruisearch42

File: python/ray/dag/tests/experimental/test_accelerated_dag.py (hunk 7 above)
Comment (Bye-legumes) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:2497]: fixed!

File: python/ray/dag/tests/experimental/test_accelerated_dag.py (hunk 8 above)
Comment (Bye-legumes) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py]: changed to check nodes and edges with doc

File: python/ray/dag/compiled_dag_node.py (hunk 9 above)
Comment (Bye-legumes) [reply python/ray/dag/compiled_dag_node.py:1901]: fixed!

File: python/ray/dag/compiled_dag_node.py (hunk 11 above)
Comment (Bye-legumes) [reply python/ray/dag/compiled_dag_node.py]: fixed

File: python/ray/dag/compiled_dag_node.py (hunk 12 above)
Comment (Bye-legumes) [reply python/ray/dag/compiled_dag_node.py:1899]: fixed!

File: python/ray/dag/tests/experimental/test_accelerated_dag.py
Comment (ruisearch42) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:1]: I think we can introduce something like a test requirement.txt , something like python/requirements/ml/data-test-requirements.txt 
If that is too complex, we can skip the test for now.

Comment (Bye-legumes) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:1]: I just skip these tests and I think it's OK now.

**Review by ruisearch42 — APPROVED** [review]
Thanks for the change!

**Comment by ruisearch42** [comment]
@Bye-legumes Auto merge is enabled, but tests are failing, please fix

**Comment by Bye-legumes** [comment]
> @Bye-legumes Auto merge is enabled, but tests are failing, please fix

I see, thx you keep the eye on this! I think this tests failed is not related as other PR also failed on same test. Let me merge the master and check if it will recurs!

**Comment by Bye-legumes** [comment]
@ruisearch42 Hi, I just merged the master and it's OK now!

**Comment by ruisearch42** [comment]
@rkooo567 to merge

**Pull request merged**